### PR TITLE
Timebased gc

### LINKED ISSUE #6128: [FR] Run `mlflow gc` based on time since last deletion
### Willingness to contribute

Yes. I would be willing to contribute this feature with guidance from the MLflow community.

### Proposal Summary

Adding more functionality to `mlflow gc`, where administrators can delete runs based on time criteria (e.g. delete runs that have been tagged as `deleted` > 1 month ago). This could be done by adding a `delete_date` column into `runs` table in the sqlstore.

### Motivation

> #### What is the use case for this feature?
Sometimes developers would want to restore their runs that they previously deleted, so just running `mlflow gc` to delete all `deleted` runs periodically might cause developers to lose their data. By being able to delete runs that have been deleted for some time, we can be more sure that the deleted run is no longer needed.

> #### Why is this use case valuable to support for MLflow users in general?
Sometimes developers would want to restore their runs that they previously deleted, so just running `mlflow gc` to delete all `deleted` runs periodically might cause developers to lose their data. By being able to delete runs that have been deleted for some time, we can be more sure that the deleted run is no longer needed.

> #### Why is this use case valuable to support for your project(s) or organization?
Sometimes developers would want to restore their runs that they previously deleted, so just running `mlflow gc` to delete all `deleted` runs periodically might cause developers to lose their data. By being able to delete runs that have been deleted for some time, we can be more sure that the deleted run is no longer needed.

> #### Why is it currently difficult to achieve this use case?
`runs` table only have `start_time` and `end_time` column that tracks when the run is created and finished. However, it does not track when the run is deleted so it is impossible to delete based on this criteria.

### Details

_No response_

### What component(s) does this bug affect?

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [X] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### PR DESCRIPTION
**Pull request by Jason-CKY**
## Related Issues/PRs

Resolves #6128 

## What changes are proposed in this pull request?

Added `delete_time` column in `Run` table to support `--older-than` flag in `mlflow gc` command. Allows to delete runs marked for deletion that has not been restored by a certain time. 

(Please fill in changes proposed in this fix)

## How is this patch tested?
Unit tests

- [x] I have written tests (not required for typo or doc fix) and confirmed the proposed feature/bug-fix/change works.

Example command to delete runs that are marked for deletion more than 30 days ago.
```sh
mlflow gc --run-ids 52705128 --older-than 30d
```
The `Runs` table has an additional column called `delete_time`, that updates on the timestamp when the run is marked for deletion. This allows the `gc` command to delete runs that are stale for a certain amount of time during routing cleaning. The column is updated to null value once the run is restored from a deleted state.

The `--older-than` flag is a string in #d#h#m#s format. For example: `mlflow gc --older-than 1d2h3m4s` Defaults to 0 (all runs). 

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Check the status of the `ci/circleci: build_doc` check. If it's successful, proceed to the
   next step, otherwise fix it.
2. Click `Details` on the right to open the job page of CircleCI.
3. Click the `Artifacts` tab.
4. Click `docs/build/html/index.html`.
5. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)
Added `--older-than` flag in `mlflow gc` command to only hard delete runs that's been marked for deletion for a period of time.

### What component(s), interfaces, languages, and integrations does this PR affect?
Components
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/pipelines`: Pipelines, Pipeline APIs, Pipeline configs, Pipeline Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [x] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/protos/service.proto
Comment (dbczumar) [inline mlflow/protos/service.proto:641]: @Jason-CKY Rather than introducing a change to the `RunInfo` entity / proto, which has a larger change footprint that includes implementation work across several third-party MLflow services, can we start by adding the `delete_time` column / attribute to the SQLAlchemyStore / FileStore without modifying the proto? In the future, we may choose to extend the proto.

Comment (Jason-CKY) [reply mlflow/protos/service.proto:641]: I understand. we can start with small changes first. I changed the proto as it was causing some of the test cases to fail, but i'll revert the changes to protos!

File: mlflow/cli.py
Comment (dbczumar) [inline mlflow/cli.py:491]: The default value of `0` doesn't seem to conform to the proposed datetime string format of `#d#h#m#s`. Can we default to `None` instead? I think that this may also improve code readability.

File: mlflow/cli.py
Comment (dbczumar) [inline mlflow/cli.py:544]: Can we introduce an optional `older_than` parameter to `_get_deleted_runs()`? This will improve efficiency for the SQLAlchemyStore, and it will also improve code readability here.

File: mlflow/entities/run_info.py
Comment (dbczumar) [inline mlflow/entities/run_info.py:96]: If we adopt the proposal from https://github.com/mlflow/mlflow/pull/6354/files#r933551129, I think we can avoid having to expose the deletion time on the RunInfo.

**Review by dbczumar — COMMENTED** [review]
@Jason-CKY This is looking great! Just left a few small comments. Looking forward to getting this merged!

File: mlflow/entities/run_info.py
Comment (Jason-CKY) [reply mlflow/entities/run_info.py:96]: The run_info is used by the file_store's `delete_run` function, where delete_time will be updated into the file store. if we remove this line how do I get the delete_time information on file_store repository?

Comment (Jason-CKY) [reply mlflow/entities/run_info.py:96]: I realised that without changes to `RunInfo` and the protos, the `--older-than` flag would not work for file-stores. Do you want this feature to only work with sqlalchemy databases?

**Comment by Jason-CKY** [comment]
I have implemented the relevant changes, regarding the `delete_time` parameter from RunInfo class, I have written the information into the yaml file directly instead of storing it there for FileStorage. Hopefully this is an acceptable workaround :)

File: mlflow/cli.py
Comment (dbczumar) [inline mlflow/cli.py:557]: Can we add `error_code=INVALID_PARAMETER_VALUE` here?

File: mlflow/store/tracking/file_store.py
Comment (dbczumar) [inline mlflow/store/tracking/file_store.py:504]: What happens if the `delete_time` key doesn't exist in the meta.yaml file?

Comment (dbczumar) [inline mlflow/store/tracking/file_store.py:504]: Nit: Can we call this `deleted_time` instead?

File: mlflow/store/tracking/file_store.py
Comment (dbczumar) [inline mlflow/store/tracking/file_store.py:942]:
```suggestion
        if delete_time is not None:
```

File: mlflow/store/tracking/sqlalchemy_store.py
Comment (dbczumar) [inline mlflow/store/tracking/sqlalchemy_store.py:482]: Nit: Can we use `deleted_time` instead?

```suggestion
        run.deleted_time = int(time.time() * 1000)
```

File: mlflow/cli.py
Comment (dbczumar) [inline mlflow/cli.py:541]: Is the copy necessary here?

File: mlflow/store/db_migrations/versions/0c779009ac13_add_delete_time_field_to_runs_table.py
Comment (dbczumar) [inline mlflow/store/db_migrations/versions/0c779009ac13_add_delete_time_field_to_runs_table.py:20]:
```suggestion
    op.add_column("runs", sa.Column("deleted_time", sa.BigInteger, nullable=True, default=None))
```

**Review by dbczumar — COMMENTED** [review]
@Jason-CKY Looks great! Just a few more small comments. Should be all set to merge once they're addressed :)

File: mlflow/store/tracking/file_store.py
Comment (Jason-CKY) [reply mlflow/store/tracking/file_store.py:504]: I'll set the condition to delete the run if `delete_time` key does not exist as well

File: mlflow/cli.py
Comment (Jason-CKY) [reply mlflow/cli.py:542]: i removed the copy and just renamed the variable here

File: mlflow/cli.py
Comment (Jason-CKY) [reply mlflow/cli.py]: Added!

File: mlflow/store/tracking/file_store.py
Comment (Jason-CKY) [reply mlflow/store/tracking/file_store.py]: :ok_hand:

File: mlflow/store/tracking/sqlalchemy_store.py
Comment (Jason-CKY) [reply mlflow/store/tracking/sqlalchemy_store.py]: :ok_hand:

File: mlflow/store/db_migrations/versions/0c779009ac13_add_delete_time_field_to_runs_table.py
Comment (Jason-CKY) [reply mlflow/store/db_migrations/versions/0c779009ac13_add_delete_time_field_to_runs_table.py]: :ok_hand:

**Comment by Jason-CKY** [comment]
I have made the corresponding changes! Thanks for the help and guidance :)

File: tests/db/schemas/mssql.sql
Comment (dbczumar) [inline tests/db/schemas/mssql.sql:72]: Nit: I think we'll need to update these schemas to reflect the rename to `deleted_time`:

```suggestion
	deleted_time BIGINT,
```

File: mlflow/cli.py
Comment (dbczumar) [inline mlflow/cli.py:492]:
```suggestion
    help="Optional. Remove run(s) older than the specified time limit. "
```

File: mlflow/cli.py
Comment (dbczumar) [inline mlflow/cli.py:535]: Can we also include `INVALID_PARAMETER_VALUE` as an error code here?

File: mlflow/cli.py
Comment (dbczumar) [inline mlflow/cli.py:526]:
```suggestion
    if older_than is not None:
```

File: mlflow/store/tracking/file_store.py
Comment (dbczumar) [inline mlflow/store/tracking/file_store.py:504]:
```suggestion
            if "deleted_time" not in meta or current_time - int(meta["deleted_time"]) > older_than:
```

**Review by dbczumar — APPROVED** [review]
@Jason-CKY LGTM with a few small comments to address before merging. Thank you so much for your valuable contribution!

File: mlflow/store/db_migrations/versions/0c779009ac13_add_delete_time_field_to_runs_table.py
Comment (harupy) [inline mlflow/store/db_migrations/versions/0c779009ac13_add_delete_time_field_to_runs_table.py:1]: Nit: Can we rename this file to `0c779009ac13_add_deleted_time_field_to_runs_table.py` (note `deleted`)?

File: tests/test_cli.py
Comment (harupy) [inline tests/test_cli.py:205]: Can we include the expected error message in `match`?

```suggestion
    with pytest.raises(subprocess.CalledProcessError, match=r"Run \w+ is not older than the required age."):
```

File: mlflow/store/tracking/sqlalchemy_store.py
Comment (harupy) [inline mlflow/store/tracking/sqlalchemy_store.py:675]: Can we filter by `deleted_time` when fetching runs from DB?

```suggestion
            runs = (
                session.query(SqlRun)
                .filter(
                    SqlRun.lifecycle_stage == LifecycleStage.DELETED,
                    SqlRun.deleted_time < (current_time - older_than),
                )
                .all()
```

File: mlflow/cli.py
Comment (harupy) [inline mlflow/cli.py:494]:
```suggestion
    "For example: --older-than 1.2d3h4m5s",
```

Can we indicate a `float` value is also supported?

File: tests/test_cli.py
Comment (harupy) [reply tests/test_cli.py:205]: The code above doesn't seem to work, but the code below does:

```python
    with pytest.raises(subprocess.CalledProcessError, match=r".+") as exp:
        subprocess.run(
            [
                "mlflow",
                "gc",
                "--backend-store-uri",
                sqlite_store[1],
                "--older-than",
                "10d10h10m10s",
                "--run-ids",
                run.info.run_uuid,
            ],
            check=True,
            capture_output=True,
            text=True,
        )
    assert "is not older than the required age" in exp.value.stderr
```

**Review by harupy — APPROVED** [review]
Left some comments. LGTM once they are addressed!

File: mlflow/cli.py
Comment (harupy) [reply mlflow/cli.py:494]:
```suggestion
    "For example: 1d2h3m4s, 1.5d, or 30s",
```

Providing multiple examples might be easier to understand.

File: tests/db/schemas/mssql.sql
Comment (Jason-CKY) [reply tests/db/schemas/mssql.sql]: 👌

File: mlflow/cli.py
Comment (Jason-CKY) [reply mlflow/cli.py]: 👌

File: mlflow/cli.py
Comment (Jason-CKY) [reply mlflow/cli.py]: 👌

File: mlflow/cli.py
Comment (Jason-CKY) [reply mlflow/cli.py]: 👌

File: mlflow/store/tracking/file_store.py
Comment (Jason-CKY) [reply mlflow/store/tracking/file_store.py]: 👌

File: mlflow/store/db_migrations/versions/0c779009ac13_add_delete_time_field_to_runs_table.py
Comment (Jason-CKY) [reply mlflow/store/db_migrations/versions/0c779009ac13_add_delete_time_field_to_runs_table.py]: 👌

File: tests/test_cli.py
Comment (Jason-CKY) [reply tests/test_cli.py]: I have replaced the code with the snippet you have provided for both gc test case with file store and sqlite store

File: mlflow/cli.py
Comment (Jason-CKY) [reply mlflow/cli.py]: 👌

File: mlflow/store/tracking/sqlalchemy_store.py
Comment (Jason-CKY) [reply mlflow/store/tracking/sqlalchemy_store.py]: 👌

**Comment by Jason-CKY** [comment]
Made another round of changes. Thanks again for the timely feedback as always!

File: tests/test_cli.py
Comment (harupy) [inline tests/test_cli.py:301]: This assert statement is ignored because `subprocess.run` throws.

```suggestion
    assert "is not older than the required age" in exp.value.stderr
```


https://docs.pytest.org/en/7.1.x/reference/reference.html#pytest.raises for more information.

image

File: tests/test_cli.py
Comment (harupy) [inline tests/test_cli.py:221]: https://github.com/mlflow/mlflow/pull/6354/files#r936803720 Same here

File: tests/resources/db/initial_models.py
Comment (harupy) [inline tests/resources/db/initial_models.py:121]:
```suggestion
    deleted_time = Column(BigInteger, nullable=True, default=None)
```

File: tests/store/tracking/test_sqlalchemy_store.py
Comment (harupy) [inline tests/store/tracking/test_sqlalchemy_store.py:287]:
```suggestion
            self.assertTrue(restored_run.deleted_time is None)
```

File: tests/store/tracking/test_sqlalchemy_store.py
Comment (harupy) [inline tests/store/tracking/test_sqlalchemy_store.py:747]:
```suggestion
            "deleted_time": int(time.time()),
```

File: tests/store/tracking/test_sqlalchemy_store.py
Comment (harupy) [inline tests/store/tracking/test_sqlalchemy_store.py:835]:
```suggestion
                actual.deleted_time is not None
```

File: tests/store/tracking/test_sqlalchemy_store.py
Comment (harupy) [inline tests/store/tracking/test_sqlalchemy_store.py:1252]:
```suggestion
        self.assertTrue(restored.info.deleted_time is None)
```

File: tests/store/tracking/test_sqlalchemy_store.py
Comment (harupy) [inline tests/store/tracking/test_sqlalchemy_store.py:2344]:
```suggestion
    assert models.SqlRun.get_attribute_name("deleted_time") == "deleted_time"
```

**Comment by harupy** [comment]
@Jason-CKY Thanks for the updates! Can you replace `delete_time` with `deleted_time`?

File: tests/store/tracking/test_sqlalchemy_store.py
Comment (harupy) [reply tests/store/tracking/test_sqlalchemy_store.py:835]: Can you also replace other `delete_time`?

**Comment by Jason-CKY** [comment]
Sorry for the repeated edits, should have changed all of the `delete_run` to `deleted_run` this time. 🙏 

I have also corrected the `test_file_store.py` and `test_sqlalchemy_store.py` files to check for `deleted_time` attribute via the meta yaml file or SqlRun instead of the RunInfo object as `deleted_time` is no longer an attribute of RunInfo

File: tests/test_cli.py
Comment (Jason-CKY) [reply tests/test_cli.py]: Thanks for the information. I have changed it such that it runs the assert statement and passed the pytest for this file!

File: tests/store/tracking/test_sqlalchemy_store.py
Comment (Jason-CKY) [reply tests/store/tracking/test_sqlalchemy_store.py]: This didn't work as `deleted_time` is no longer an attribute of `RunInfo`. I have changed it to query for `SqlRun` for the `deleted_time` attribute instead

**Comment by dbczumar** [comment]
> Sorry for the repeated edits, should have changed all of the `delete_run` to `deleted_run` this time. 🙏
> 
> I have also corrected the `test_file_store.py` and `test_sqlalchemy_store.py` files to check for `deleted_time` attribute via the meta yaml file or SqlRun instead of the RunInfo object as `deleted_time` is no longer an attribute of RunInfo

No problem at all! Looks great! Happy to merge once conflicts with the `master` branch are resolved.

File: tests/store/tracking/test_file_store.py
Comment (harupy) [inline tests/store/tracking/test_file_store.py:529]:
```suggestion
        assert "deleted_time" not in meta or meta["deleted_time"] is None
```

Do we need `or meta["deleted_time"] is None`?

Comment (Jason-CKY) [reply tests/store/tracking/test_file_store.py]: you're right, i've removed the `meta["deleted_time"] is None` check!

**Comment by harupy** [comment]
@Jason-CKY Can you resolve conflicts?

**Comment by Jason-CKY** [comment]
Conflicts resolved!

File: mlflow/cli.py
Comment (harupy) [inline mlflow/cli.py:15]: Can we revert this change?

File: mlflow/store/tracking/file_store.py
Comment (harupy) [inline mlflow/store/tracking/file_store.py:3]: Let's remove this.

**Comment by harupy** [comment]
@Jason-CKY Thanks for the update! Can you fix the lint errors?

File: mlflow/cli.py
Comment (Jason-CKY) [reply mlflow/cli.py]: :ok_hand:

File: mlflow/store/tracking/file_store.py
Comment (Jason-CKY) [reply mlflow/store/tracking/file_store.py:3]: :ok_hand:

**Review by harupy — APPROVED** [review]

**Comment by Jason-CKY** [comment]
fixed lint errors! thanks for checking

**Comment by harupy** [comment]
@Jason-CKY Let me push a commit to fix the python check failure.

**Comment by harupy** [comment]
@Jason-CKY Merged! Thanks for the contribution!

**Pull request merged**
